geoh5py/shared/entity.py

```diff
@@ -170,6 +170,22 @@ def create(cls, workspace, **kwargs):
 
         return new_object
 
+    def delete(self):
+        """
+        Function to delete itself and children recursively from the workspace
+        """
+        for child in self.children:
+            child.delete()
+
+        self.parent.children.remove(self)
+
+        if getattr(self.parent, "property_groups", None) is not None:
+            for prop_group in self.parent.property_groups:
+                if self.uid in prop_group.properties:
+                    self.parent.remove_data_from_group(self.uid, prop_group.name)
+
+        self.workspace.delete_entity(self)
```

Comment:
**thought:** logic seems backward. I would not expect that call to workspace here

Could instead the following work?
Entity is removed from workspace, which lead to its deletion (cascading to children, etc...)
When Entity is deleted, the weakref become null in the workspace.
Thus after the cascade deletion process, the workspace iterates on the various dict[uid, weakref], and for any null weakref notify the writer to delete the corresponding uid.

Reply:
forget about my comment above and weakref.

maybe it is just about renaming `delete_entity` in workspace
It just feels wrong to have 2 delete functions,

If the action starts with some workspace.remove_entity(entity)
then workspace could recursively get the children and do all the cleanup, and update with the writer.

To be discussed...

Reply:
I implemented part of your first idea:
1- The workspace removes the entity from the list of children from the parent (which brakes the weakref of the entity, and entity_type is not used elsewhere)
2- Workspace calls H5writer to remove the object on file.
3- At the end of the recursive remove of entity and children, we force garbage collection and check weakrefs for all entity_types. If weakref is None, then H5writer is called and removed from geoh5.